### PR TITLE
Fix clippy lints

### PR DESCRIPTION
**Pull request by dtolnay**
Clippy was not running because the Travis build used the "clippy" feature and clippy was cfg'd with the "nightly-testing" feature.

### CODE REVIEW

File: json/src/value.rs [hunk 1]
```diff
@@ -468,10 +468,14 @@ struct WriterFormatter<'a, 'b: 'a> {
...
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        match self.inner.write_str(str::from_utf8(buf).unwrap()) {
-            Ok(_) => Ok(buf.len()),
-            Err(_) => Err(io::Error::last_os_error()),
+        fn io_error<E>(_: E) -> io::Error {
+            // Value does not matter because fmt::Debug and fmt::Display impls
+            // below just map it to fmt::Error
+            io::Error::new(io::ErrorKind::Other, "fmt error")
         }
+        let s = try!(str::from_utf8(buf).map_err(io_error));
```
Comment (oli-obk) [inline json/src/value.rs]: nit: this could be `.or(io_error())` (+ remove the argument from `io_error`)

File: json/src/value.rs [hunk 2]
```diff
@@ -468,10 +468,14 @@ struct WriterFormatter<'a, 'b: 'a> {
 
 impl<'a, 'b> io::Write for WriterFormatter<'a, 'b> {
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        match self.inner.write_str(str::from_utf8(buf).unwrap()) {
-            Ok(_) => Ok(buf.len()),
-            Err(_) => Err(io::Error::last_os_error()),
```
Comment (oli-obk) [inline json/src/value.rs]: the semantics changed, isn't the `last_os_error` important?

**Comment by oli-obk** [comment]
do we need to change something in travis, too?

**Comment by oli-obk** [comment]
Great cleanup. Can be merged as is if the comments are irrelevant.

File: json/src/value.rs (hunk 1 above)
Comment (dtolnay) [inline json/src/value.rs]: In `.or(io_error())` the argument is evaluated even if there is no error. And `io::Error::new` is [actually pretty expensive](https://github.com/rust-lang/rust/blob/de78655bca47cac8e783dbb563e7e5c25c1fae40/src/libstd/io/error.rs#L191-L204) - two memory allocations, one on the `into()` call that makes a Box, and another explicit `Box::new`. So this function would end up doing 4 useless memory allocations on every call to `write`... which may itself by called many many times as a value is being serialized.

File: json/src/value.rs (hunk 2 above)
Comment (dtolnay) [inline json/src/value.rs]: The only callers are the `fmt::Debug` and `fmt::Display` impls below, which immediately do `.map_err(|_| fmt::Error)` to throw away the error. I mention this in the comment inside `io_error`.

Also I doubt that `fmt::Formatter::write_str` or `str::from_utf8` would set `errno` anyway, right?

**Comment by dtolnay** [comment]
> do we need to change something in travis, too?

No, I changed lib.rs so clippy runs in Travis now.

```diff
- #![cfg_attr(feature = "nightly-testing", plugin(clippy))]
+ #![cfg_attr(feature = "clippy", plugin(clippy))]
```

**Comment by dtolnay** [comment]
Travis passes `--features nightly-testing` to json_test [here](https://github.com/serde-rs/json/blob/bfd835f2023ad0f2e789892f4151fac6ae7d236b/.travis.yml#L33) and json_test passes `--features clippy` to serde_json [here](https://github.com/serde-rs/json/blob/bfd835f2023ad0f2e789892f4151fac6ae7d236b/json_tests/Cargo.toml#L10).

**Pull request merged**